### PR TITLE
WIP: Cpp Export

### PR DESCRIPTION
**Pull request by patrick-luethi**
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45758692%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45759048%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45759271%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45759547%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45759710%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45761641%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45761819%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45762642%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45762965%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45763011%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/exporter/CppExporter.cpp%20153%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45761641%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Please%20move%20this%20to%20%60CodeComposite%60%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A49%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/exporter/CppExporter.cpp%3AL34-191%22%7D%2C%20%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/exporter/CppExporter.cpp%20145%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45761819%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Like%20the%20previous%20method%2C%20this%20loop%20should%20just%20call%20%60CodeComposite%3A%3AcalculateDependencies%60%20which%20in%20turn%20should%20call%20%60CodeUnit%3A%3AcalculateDependencies%60.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A50%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/exporter/CppExporter.cpp%3AL34-191%22%7D%2C%20%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/exporter/CppExporter.cpp%20119%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45759547%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20think%20this%20inner%20loop%20should%20not%20be%20in%20%60CppExport%60%20but%20rather%20in%20each%20%60Composite%60.%20So%20a%20composite%20should%20have%20a%20%60calculateSourceFragments%60%20method.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A36%3A06Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/exporter/CppExporter.cpp%3AL34-191%22%7D%2C%20%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/exporter/CppExporter.cpp%20179%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45762642%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Please%20deduplicate%20the%20stuff%20below%2C%20it%27s%20identical%20to%20the%20stuff%20above.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A56%3A26Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/exporter/CppExporter.cpp%3AL34-191%22%7D%2C%20%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/exporter/CppExporter.cpp%2016%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45758692%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Please%20do%20not%20use%20names%20such%20as%20%60us%60%20and%20%60cs%60.%20Use%20proper%20descriptive%20names.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A30%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/exporter/CppExporter.cpp%3AL34-191%22%7D%2C%20%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/visitors/source_visitors/ExpressionVisitorSource.cpp%2011%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45763011%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Please%20put%20this%20on%20the%20previous%20line%2C%20and%20also%20the%203%20similar%20things%20below.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A58%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/visitors/source_visitors/ExpressionVisitorSource.cpp%3AL186-208%22%7D%2C%20%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/exporter/CppExporter.cpp%20120%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45759710%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Also%20the%20loop%20in%20%60calculateSourceFragments%60%20should%20be%20nothing%20more%20than%20a%20call%20to%20a%20%60calculateSourceFragments%60%20method%20inside%20each%20Unit%20which%20contains%20this%20if%20statement.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A37%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/exporter/CppExporter.cpp%3AL34-191%22%7D%2C%20%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/visitors/header_visitors/ExpressionVisitorHeader.cpp%2011%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45762965%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Please%20put%20this%20on%20the%20previous%20line%2C%20also%20the%20same%20for%20all%203%20things%20below.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A58%3A20Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/visitors/header_visitors/ExpressionVisitorHeader.cpp%3AL186-208%22%7D%2C%20%22Pull%2050dd167395983720973d7c94723b7e37c7135ca8%20CppExport/src/exporter/CppExporter.cpp%2032%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/222%23discussion_r45759048%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Do%20we%20need%20to%20have%20two%20%60units%60%20method%3F%20Is%20the%20other%20method%20called%20anywhere%20else%20but%20below%3F%20If%20so%20%28or%20if%20you%20expect%20this%20to%20happen%29%2C%20where%20and%20why%3F%20If%20not%2C%20then%20please%20merge%20these%20into%20one%20method.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A32%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20can%20just%20remove%20the%20wrapper%20method.%22%2C%20%22created_at%22%3A%20%222015-11-24T16%3A34%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6049349%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/patrick-luethi%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20CppExport/src/exporter/CppExporter.cpp%3AL34-191%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/exporter/CppExporter.cpp 16'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45758692'>File: CppExport/src/exporter/CppExporter.cpp:L34-191</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Please do not use names such as `us` and `cs`. Use proper descriptive names.
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/exporter/CppExporter.cpp 32'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45759048'>File: CppExport/src/exporter/CppExporter.cpp:L34-191</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Do we need to have two `units` method? Is the other method called anywhere else but below? If so (or if you expect this to happen), where and why? If not, then please merge these into one method.
- <a href='https://github.com/patrick-luethi'></a> I can just remove the wrapper method.
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/exporter/CppExporter.cpp 119'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45759547'>File: CppExport/src/exporter/CppExporter.cpp:L34-191</a></b>
- <a href='https://github.com/dimitar-asenov'></a> I think this inner loop should not be in `CppExport` but rather in each `Composite`. So a composite should have a `calculateSourceFragments` method.
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/exporter/CppExporter.cpp 120'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45759710'>File: CppExport/src/exporter/CppExporter.cpp:L34-191</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Also the loop in `calculateSourceFragments` should be nothing more than a call to a `calculateSourceFragments` method inside each Unit which contains this if statement.
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/exporter/CppExporter.cpp 153'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45761641'>File: CppExport/src/exporter/CppExporter.cpp:L34-191</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Please move this to `CodeComposite`
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/exporter/CppExporter.cpp 145'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45761819'>File: CppExport/src/exporter/CppExporter.cpp:L34-191</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Like the previous method, this loop should just call `CodeComposite::calculateDependencies` which in turn should call `CodeUnit::calculateDependencies`.
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/exporter/CppExporter.cpp 179'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45762642'>File: CppExport/src/exporter/CppExporter.cpp:L34-191</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Please deduplicate the stuff below, it's identical to the stuff above.
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/visitors/header_visitors/ExpressionVisitorHeader.cpp 11'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45762965'>File: CppExport/src/visitors/header_visitors/ExpressionVisitorHeader.cpp:L186-208</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Please put this on the previous line, also the same for all 3 things below.
- [x] <a href='#crh-comment-Pull 50dd167395983720973d7c94723b7e37c7135ca8 CppExport/src/visitors/source_visitors/ExpressionVisitorSource.cpp 11'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/222#discussion_r45763011'>File: CppExport/src/visitors/source_visitors/ExpressionVisitorSource.cpp:L186-208</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Please put this on the previous line, and also the 3 similar things below.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/222?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/222?approve=1'></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/222'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by dimitar-asenov** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/222'>Refresh</a>.<a href='#crh-update-none'></a>

File: CppExport/src/exporter/CppExporter.cpp [hunk 1]
```diff
@@ -34,74 +34,157 @@
...
+#include "../CodeComposite.h"
+#include "../Config.h"
 
 namespace CppExport {
 
 QList<ExportError> CppExporter::exportTree(Model::TreeManager* manager, const QString& pathToProjectContainerDirectory)
 {
-	auto project = DCast<OOModel::Project>(manager->root());
-	Q_ASSERT(project);
+	auto us = units(manager);
```
Comment (dimitar-asenov) [inline CppExport/src/exporter/CppExporter.cpp]: Please do not use names such as `us` and `cs`. Use proper descriptive names.

File: CppExport/src/exporter/CppExporter.cpp [hunk 2]
```diff
@@ -34,74 +34,157 @@
...
+
+	auto dir = new Export::SourceDir(nullptr, pathToProjectContainerDirectory + "/src");
+	for (auto c : cs) exportCodeComposite(c, dir);
+	auto layout = layouter();
+	Export::Exporter::exportToFileSystem("", dir, &layout);
+
+	return {};
+}
+
+QList<CodeUnit*> CppExporter::units(Model::TreeManager* treeManager)
```
Comment (dimitar-asenov) [inline CppExport/src/exporter/CppExporter.cpp]: Do we need to have two `units` method? Is the other method called anywhere else but below? If so (or if you expect this to happen), where and why? If not, then please merge these into one method.

Comment (patrick-luethi) [inline CppExport/src/exporter/CppExporter.cpp]: I can just remove the wrapper method.

File: CppExport/src/exporter/CppExporter.cpp [hunk 3]
```diff
@@ -34,74 +34,157 @@
...
-	auto file = &dir->file("exported.h");
-	file->append(fragment);
-	auto layout = layouter();
-	Export::Exporter::exportToFileSystem("", dir, &layout);
+	DeclarationVisitorHeader headerVisitor;
+	DeclarationVisitorSource sourceVisitor;
+
+	for (auto codeComposite : codeComposites)
+	{
+		for (auto unit : codeComposite->units())
```
Comment (dimitar-asenov) [inline CppExport/src/exporter/CppExporter.cpp]: I think this inner loop should not be in `CppExport` but rather in each `Composite`. So a composite should have a `calculateSourceFragments` method.

File: CppExport/src/exporter/CppExporter.cpp [hunk 4]
```diff
@@ -34,74 +34,157 @@
...
+			if (auto classs = DCast<OOModel::Class>(unit->node()))
```
Comment (dimitar-asenov) [inline CppExport/src/exporter/CppExporter.cpp]: Also the loop in `calculateSourceFragments` should be nothing more than a call to a `calculateSourceFragments` method inside each Unit which contains this if statement.

File: CppExport/src/exporter/CppExporter.cpp [hunk 5]
```diff
@@ -34,74 +34,157 @@
...
+	for (auto codeComposite : codeComposites)
+		for (CodeUnit* unit : codeComposite->units())
+		{
+			unit->headerPart()->calculateDependencies(allHeaderParts);
+			unit->sourcePart()->calculateDependencies(allHeaderParts);
+		}
 }
 
-void CppExporter::exportCodeComposite(CodeComposite* codeComposite)
+void CppExporter::exportCodeComposite(CodeComposite* codeComposite, Export::SourceDir* dir)
```
Comment (dimitar-asenov) [inline CppExport/src/exporter/CppExporter.cpp]: Please move this to `CodeComposite`

File: CppExport/src/exporter/CppExporter.cpp [hunk 6]
```diff
@@ -34,74 +34,157 @@
...
-			unit->setSourceFragment(visitor.visitTopLevelClass(classs));
-		else
-			Q_ASSERT(false);
+	QList<CodeUnitPart*> allHeaderParts;
+	for (auto codeComposite : codeComposites)
+		for (auto unit : codeComposite->units())
+			allHeaderParts.append(unit->headerPart());
+
+	for (auto codeComposite : codeComposites)
+		for (CodeUnit* unit : codeComposite->units())
```
Comment (dimitar-asenov) [inline CppExport/src/exporter/CppExporter.cpp]: Like the previous method, this loop should just call `CodeComposite::calculateDependencies` which in turn should call `CodeUnit::calculateDependencies`.

File: CppExport/src/exporter/CppExporter.cpp [hunk 7]
```diff
@@ -34,74 +34,157 @@
...
 
-	auto composite = new Export::CompositeFragment(new Model::Integer());
-	for (auto unit : codeComposite->units())
-		composite->append(unit->sourceFragment());
+		auto file = &dir->file(codeComposite->name() + ".h");
+		file->append(composite);
+	}
 
-	exportFragment(composite);
+	// output source file
```
Comment (dimitar-asenov) [inline CppExport/src/exporter/CppExporter.cpp]: Please deduplicate the stuff below, it's identical to the stuff above.

File: CppExport/src/visitors/header_visitors/ExpressionVisitorHeader.cpp [hunk 8]
```diff
@@ -186,6 +186,22 @@ SourceFragment* ExpressionVisitorHeader::visit(Expression* expression)
...
 	else if (auto e = DCast<MethodCallExpression>(expression))
 		*fragment << visit(e->callee()) << list(e->arguments(), this, "argsList");
+	else if (auto e = DCast<MetaCallExpression>(expression))
+	{
+		*fragment << visit(e->callee());
+
+		auto arguments = new Export::CompositeFragment(e->arguments(), "argsList");
+		for (auto node : *e->arguments())
+			if (auto n = DCast<Expression>(node))
+				*arguments << visit(n);
```
Comment (dimitar-asenov) [inline CppExport/src/visitors/header_visitors/ExpressionVisitorHeader.cpp]: Please put this on the previous line, also the same for all 3 things below.

File: CppExport/src/visitors/source_visitors/ExpressionVisitorSource.cpp [hunk 9]
```diff
@@ -186,6 +186,22 @@ SourceFragment* ExpressionVisitorSource::visit(Expression* expression)
...
 	else if (auto e = DCast<MethodCallExpression>(expression))
 		*fragment << visit(e->callee()) << list(e->arguments(), this, "argsList");
+	else if (auto e = DCast<MetaCallExpression>(expression))
+	{
+		*fragment << visit(e->callee());
+
+		auto arguments = new Export::CompositeFragment(e->arguments(), "argsList");
+		for (auto node : *e->arguments())
+			if (auto n = DCast<Expression>(node))
+				*arguments << visit(n);
```
Comment (dimitar-asenov) [inline CppExport/src/visitors/source_visitors/ExpressionVisitorSource.cpp]: Please put this on the previous line, and also the 3 similar things below.

**Pull request merged**